File: Autonomous Mobile Robots/Chapter15/Lesson5/Chapter15_Lesson5.py
# Chapter15_Lesson5.py
# Lab: Compare Local Planners in Dense Obstacles (DWA vs simplified TEB-proxy)
# Requirements: Python 3.9+, numpy, matplotlib (optional for plotting)

import math
import numpy as np

# ---------- World / geometry ----------
def corridor_y(x: float) -> float:
    return 0.9 * math.sin(0.6 * x)

def reference_path(n=120, x0=0.5, x1=11.5):
    xs = np.linspace(x0, x1, n)
    ys = np.array([corridor_y(x) for x in xs])
    return np.c_[xs, ys]

def sample_dense_world(n_obs=45, seed=0):
    rng = np.random.default_rng(seed)
    obs = []
    xmin, xmax, ymin, ymax = 0.0, 12.0, -3.0, 3.0
    tries = 0
    while len(obs) < n_obs and tries < 20000:
        tries += 1
        x = rng.uniform(xmin + 0.5, xmax - 0.5)
        y = rng.uniform(ymin + 0.5, ymax - 0.5)
        r = rng.uniform(0.12, 0.32)
        if abs(y - corridor_y(x)) < 0.55 + r:
            continue
        ok = True
        for (ox, oy, orad) in obs:
            if (x - ox) ** 2 + (y - oy) ** 2 < (r + orad + 0.05) ** 2:
                ok = False
                break
        if ok:
            obs.append((x, y, r))
    return np.array(obs, dtype=float)

def clearance(p, obs, R):
    d = np.inf
    for (ox, oy, r) in obs:
        d = min(d, math.hypot(p[0] - ox, p[1] - oy) - r - R)
    return d

def wrap(a):
    return (a + math.pi) % (2 * math.pi) - math.pi

# ---------- Unicycle ----------
def step_unicycle(x, v, w, dt):
    return np.array([x[0] + v * math.cos(x[2]) * dt,
                     x[1] + v * math.sin(x[2]) * dt,
                     wrap(x[2] + w * dt)], float)

# ---------- DWA (short) ----------
def rollout_min_clear(x, v, w, obs, R, dt, T):
    steps = max(1, int(round(T / dt)))
    xc = x.copy()
    mc = np.inf
    for _ in range(steps):
        xc = step_unicycle(xc, v, w, dt)
        mc = min(mc, clearance(xc[:2], obs, R))
        if mc <= 0:
            break
    return xc, mc

def dwa_command(x, v0, w0, path, obs, prm):
    dt, T = prm["dt"], prm["T"]
    vmin = max(0.0, v0 - prm["a"] * dt); vmax = min(prm["vmax"], v0 + prm["a"] * dt)
    wmin = max(-prm["wmax"], w0 - prm["alpha"] * dt); wmax = min(prm["wmax"], w0 + prm["alpha"] * dt)

    best = (-1e18, 0.0, 0.0)
    goal = path[-1]

    for v in np.linspace(vmin, vmax, 9):
        for w in np.linspace(wmin, wmax, 17):
            xf, mc = rollout_min_clear(x, v, w, obs, prm["R"], dt, T)
            if mc <= prm["R"] + 0.05:
                continue
            goal_dist = np.linalg.norm(xf[:2] - goal)
            J = -0.6 * goal_dist + 1.8 * mc + 0.4 * (v / (prm["vmax"] + 1e-9)) - 0.12 * (w * w)
            if J > best[0]:
                best = (J, float(v), float(w))
    return best[1], best[2]

# ---------- TEB-proxy (short band optimization) ----------
def optimize_band(x, path, obs, prm, N=12, band_len=2.6, iters=20, step=0.12):
    # sample N points ahead along x (good enough for corridor reference)
    x0 = x[0]
    xs = np.linspace(x0, x0 + band_len, N)
    pts = np.c_[xs, np.array([corridor_y(xx) for xx in xs])]
    pts[0] = x[:2]

    for _ in range(iters):
        g = np.zeros_like(pts)

        # smoothness (2nd diff)
        g[1:-1] += 0.35 * (2 * pts[1:-1] - pts[:-2] - pts[2:])

        # obstacle repulsion (nearest obstacle)
        for i in range(1, N):
            best_d = np.inf; best_u = np.zeros(2)
            for (ox, oy, r) in obs:
                dvec = pts[i] - np.array([ox, oy])
                n = np.linalg.norm(dvec) + 1e-12
                d = n - r - prm["R"]
                if d < best_d:
                    best_d = d; best_u = dvec / n
            if best_d < 0.9:
                phi = math.exp(-4.5 * (best_d - 0.9))
                g[i] += 1.8 * (-4.5 * phi) * best_u

        # time preference (shorten segments)
        g[1:] += 0.25 * (pts[1:] - pts[:-1])

        pts[1:] -= step * g[1:]
        pts[0] = x[:2]
    return pts

def teb_command(x, v0, w0, path, obs, prm):
    pts = optimize_band(x, path, obs, prm)
    d = pts[1] - pts[0]
    heading = math.atan2(d[1], d[0])
    herr = wrap(heading - x[2])
    w = max(-prm["wmax"], min(prm["wmax"], 2.2 * herr))
    v = max(0.0, min(prm["vmax"], 0.8 * (np.linalg.norm(d) / prm["dt"])))
    v *= 1.0 / (1.0 + 1.2 * abs(w))
    # accel limits
    v = max(max(0.0, v0 - prm["a"] * prm["dt"]), min(prm["vmax"], v0 + prm["a"] * prm["dt"], v))
    w = max(w0 - prm["alpha"] * prm["dt"], min(w0 + prm["alpha"] * prm["dt"], w))
    return v, w

# ---------- Benchmark ----------
def run_one(planner, seed, prm):
    obs = sample_dense_world(prm["n_obs"], seed)
    path = reference_path()
    x = np.array([0.6, 0.0, 0.0], float)
    v = 0.0; w = 0.0

    L = 0.0; cmin = np.inf; w2 = 0.0
    goal = path[-1]

    for k in range(prm["max_steps"]):
        c = clearance(x[:2], obs, prm["R"])
        cmin = min(cmin, c)
        if c <= 0:
            return dict(success=False, collision=True, time=k * prm["dt"], L=L, cmin=cmin, w2=w2)
        if np.linalg.norm(x[:2] - goal) <= prm["goal_tol"]:
            return dict(success=True, collision=False, time=k * prm["dt"], L=L, cmin=cmin, w2=w2)

        if planner == "teb":
            vcmd, wcmd = teb_command(x, v, w, path, obs, prm)
        else:
            vcmd, wcmd = dwa_command(x, v, w, path, obs, prm)

        x2 = step_unicycle(x, vcmd, wcmd, prm["dt"])
        L += float(np.linalg.norm(x2[:2] - x[:2]))
        w2 += float((wcmd ** 2) * prm["dt"])
        x, v, w = x2, vcmd, wcmd

    return dict(success=False, collision=False, time=prm["max_steps"] * prm["dt"], L=L, cmin=cmin, w2=w2)

def summarize(results):
    n = len(results)
    succ = sum(r["success"] for r in results)
    col  = sum(r["collision"] for r in results)
    def mean(key): return float(np.mean([r[key] for r in results]))
    return dict(trials=n, success_rate=succ / n, collision_rate=col / n,
                time_mean=mean("time"), L_mean=mean("L"), cmin_mean=mean("cmin"), w2_mean=mean("w2"))

def compare(trials=30, seed0=0):
    prm = dict(R=0.25, vmax=0.9, wmax=1.6, a=1.2, alpha=2.5,
               dt=0.1, T=2.0, goal_tol=0.25, max_steps=800, n_obs=45)
    for name in ["dwa", "teb"]:
        res = [run_one(name, seed0 + i, prm) for i in range(trials)]
        print(name.upper(), summarize(res))

if __name__ == "__main__":
    compare(30, 0)

File: Autonomous Mobile Robots/Chapter15/Lesson5/Chapter15_Lesson5.m
% Chapter15_Lesson5.m
% Lab: Compare Local Planners in Dense Obstacles (DWA vs simplified TEB-proxy)
% Run:
%   ComparePlanners(30, 0);

function ComparePlanners(trials, seed0)
  if nargin < 1, trials = 30; end
  if nargin < 2, seed0 = 0; end

  prm.R=0.25; prm.vmax=0.9; prm.wmax=1.6; prm.a=1.2; prm.alpha=2.5;
  prm.dt=0.1; prm.T=2.0; prm.goal_tol=0.25; prm.max_steps=800; prm.n_obs=45;

  fprintf('\nDWA: '); disp(Summarize(RunMany('dwa', trials, seed0, prm)));
  fprintf('\nTEB: '); disp(Summarize(RunMany('teb', trials, seed0, prm)));
end

function R = RunMany(name, trials, seed0, prm)
  R = repmat(struct('success',false,'collision',false,'time',0,'L',0,'cmin',inf,'w2',0), trials, 1);
  for i=1:trials
    R(i) = RunOne(name, seed0 + (i-1), prm);
  end
end

function S = Summarize(R)
  n = numel(R);
  succ = sum([R.success]); col = sum([R.collision]);
  S.trials = n;
  S.success_rate = succ/n;
  S.collision_rate = col/n;
  S.time_mean = mean([R.time]);
  S.L_mean = mean([R.L]);
  S.cmin_mean = mean([R.cmin]);
  S.w2_mean = mean([R.w2]);
end

% ---------- World / geometry ----------
function y = corridor_y(x), y = 0.9*sin(0.6*x); end

function path = reference_path()
  xs = linspace(0.5, 11.5, 120)';
  ys = arrayfun(@corridor_y, xs);
  path = [xs, ys];
end

function obs = sample_dense_world(n_obs, seed)
  rng(seed);
  obs = zeros(0,3);
  tries = 0;
  while size(obs,1) < n_obs && tries < 20000
    tries = tries + 1;
    x = 0.5 + 11.0*rand();
    y = -2.5 + 5.0*rand();
    r = 0.12 + 0.20*rand();
    if abs(y - corridor_y(x)) < 0.55 + r, continue; end
    ok = true;
    for k=1:size(obs,1)
      dx = x-obs(k,1); dy = y-obs(k,2);
      rr = r + obs(k,3) + 0.05;
      if dx*dx + dy*dy < rr*rr, ok=false; break; end
    end
    if ok, obs(end+1,:) = [x,y,r]; end %#ok<AGROW>
  end
end

function c = clearance(p, obs, R)
  d = inf;
  for i=1:size(obs,1)
    d = min(d, hypot(p(1)-obs(i,1), p(2)-obs(i,2)) - obs(i,3) - R);
  end
  c = d;
end

function a = wrap(a), a = mod(a+pi, 2*pi) - pi; end

% ---------- Unicycle ----------
function x2 = step_unicycle(x, v, w, dt)
  x2 = [x(1) + v*cos(x(3))*dt;
        x(2) + v*sin(x(3))*dt;
        wrap(x(3) + w*dt)];
end

% ---------- DWA ----------
function [xf, mc] = rollout_min_clear(x, v, w, obs, prm)
  steps = max(1, round(prm.T/prm.dt));
  mc = inf; xf = x;
  for k=1:steps
    xf = step_unicycle(xf, v, w, prm.dt);
    mc = min(mc, clearance(xf(1:2)', obs, prm.R));
    if mc <= 0, break; end
  end
end

function [vbest, wbest] = dwa_command(x, v0, w0, path, obs, prm)
  vmin = max(0, v0 - prm.a*prm.dt); vmax = min(prm.vmax, v0 + prm.a*prm.dt);
  wmin = max(-prm.wmax, w0 - prm.alpha*prm.dt); wmax = min(prm.wmax, w0 + prm.alpha*prm.dt);

  goal = path(end,:);
  bestJ = -1e18; vbest = 0; wbest = 0;

  for v = linspace(vmin, vmax, 9)
    for w = linspace(wmin, wmax, 17)
      [xf, mc] = rollout_min_clear(x, v, w, obs, prm);
      if mc <= prm.R + 0.05, continue; end
      goalDist = norm(xf(1:2)' - goal);
      J = -0.6*goalDist + 1.8*mc + 0.4*(v/(prm.vmax+1e-9)) - 0.12*w*w;
      if J > bestJ, bestJ = J; vbest = v; wbest = w; end
    end
  end
end

% ---------- TEB-proxy ----------
function pts = optimize_band(x, obs, prm, N, bandLen, iters, step)
  if nargin < 4, N=12; end
  if nargin < 5, bandLen=2.6; end
  if nargin < 6, iters=20; end
  if nargin < 7, step=0.12; end

  xs = linspace(x(1), x(1)+bandLen, N)';
  pts = [xs, arrayfun(@corridor_y, xs)];
  pts(1,:) = x(1:2)';

  for it=1:iters
    g = zeros(size(pts));

    g(2:end-1,:) = g(2:end-1,:) + 0.35*(2*pts(2:end-1,:) - pts(1:end-2,:) - pts(3:end,:));

    for i=2:N
      bestD = inf; bestU = [0,0];
      for k=1:size(obs,1)
        dvec = pts(i,:) - obs(k,1:2);
        n = norm(dvec) + 1e-12;
        d = n - obs(k,3) - prm.R;
        if d < bestD, bestD = d; bestU = dvec/n; end
      end
      if bestD < 0.9
        phi = exp(-4.5*(bestD - 0.9));
        g(i,:) = g(i,:) + 1.8*(-4.5*phi)*bestU;
      end
    end

    g(2:end,:) = g(2:end,:) + 0.25*(pts(2:end,:) - pts(1:end-1,:));

    pts(2:end,:) = pts(2:end,:) - step*g(2:end,:);
    pts(1,:) = x(1:2)';
  end
end

function [vcmd, wcmd] = teb_command(x, v0, w0, path, obs, prm) %#ok<INUSD>
  pts = optimize_band(x, obs, prm);
  d = pts(2,:) - pts(1,:);
  heading = atan2(d(2), d(1));
  herr = wrap(heading - x(3));

  wcmd = max(-prm.wmax, min(prm.wmax, 2.2*herr));
  vcmd = max(0, min(prm.vmax, 0.8*(norm(d)/prm.dt)));
  vcmd = vcmd / (1 + 1.2*abs(wcmd));

  vcmd = max(max(0, v0 - prm.a*prm.dt), min(min(prm.vmax, v0 + prm.a*prm.dt), vcmd));
  wcmd = max(w0 - prm.alpha*prm.dt, min(w0 + prm.alpha*prm.dt, wcmd));
end

% ---------- Simulation ----------
function r = RunOne(planner, seed, prm)
  obs = sample_dense_world(prm.n_obs, seed);
  path = reference_path();
  goal = path(end,:);

  x = [0.6;0.0;0.0]; v=0; w=0;
  L=0; cmin=inf; w2=0;

  for k=1:prm.max_steps
    c = clearance(x(1:2)', obs, prm.R);
    cmin = min(cmin, c);
    if c <= 0
      r = struct('success',false,'collision',true,'time',(k-1)*prm.dt,'L',L,'cmin',cmin,'w2',w2);
      return;
    end
    if norm(x(1:2)' - goal) <= prm.goal_tol
      r = struct('success',true,'collision',false,'time',(k-1)*prm.dt,'L',L,'cmin',cmin,'w2',w2);
      return;
    end

    if strcmpi(planner,'teb')
      [vcmd,wcmd] = teb_command(x,v,w,path,obs,prm);
    else
      [vcmd,wcmd] = dwa_command(x,v,w,path,obs,prm);
    end

    x2 = step_unicycle(x, vcmd, wcmd, prm.dt);
    L = L + norm(x2(1:2)-x(1:2));
    w2 = w2 + (wcmd*wcmd)*prm.dt;
    x = x2; v=vcmd; w=wcmd;
  end

  r = struct('success',false,'collision',false,'time',prm.max_steps*prm.dt,'L',L,'cmin',cmin,'w2',w2);
end

% ---------------- Optional: Simulink Harness Builder ----------------
% BuildSimulinkHarness('AMR_UnicycleHarness');

function BuildSimulinkHarness(modelName)
  if nargin < 1, modelName = 'AMR_UnicycleHarness'; end
  if bdIsLoaded(modelName), close_system(modelName,0); end
  new_system(modelName); open_system(modelName);
  set_param(modelName, 'StopTime', '20');

  add_block('simulink/Sources/From Workspace', [modelName '/Vin'], 'VariableName', 'Vin', 'Position', [30 60 140 90]);
  add_block('simulink/Sources/From Workspace', [modelName '/Win'], 'VariableName', 'Win', 'Position', [30 140 140 170]);

  add_block('simulink/Math Operations/Trigonometric Function', [modelName '/cos'], 'Operator', 'cos', 'Position', [220 40 270 80]);
  add_block('simulink/Math Operations/Trigonometric Function', [modelName '/sin'], 'Operator', 'sin', 'Position', [220 110 270 150]);
  add_block('simulink/Math Operations/Product', [modelName '/v*cos'], 'Position', [320 55 360 85]);
  add_block('simulink/Math Operations/Product', [modelName '/v*sin'], 'Position', [320 125 360 155]);

  add_block('simulink/Continuous/Integrator', [modelName '/x'], 'Position', [420 55 450 85]);
  add_block('simulink/Continuous/Integrator', [modelName '/y'], 'Position', [420 125 450 155]);
  add_block('simulink/Continuous/Integrator', [modelName '/theta'], 'Position', [220 200 250 230]);

  add_block('simulink/Sinks/To Workspace', [modelName '/x_out'], 'VariableName', 'x_out', 'SaveFormat', 'Array', 'Position', [520 55 600 85]);
  add_block('simulink/Sinks/To Workspace', [modelName '/y_out'], 'VariableName', 'y_out', 'SaveFormat', 'Array', 'Position', [520 125 600 155]);
  add_block('simulink/Sinks/To Workspace', [modelName '/theta_out'], 'VariableName', 'theta_out', 'SaveFormat', 'Array', 'Position', [320 200 410 230]);

  add_line(modelName, 'theta/1', 'cos/1'); add_line(modelName, 'theta/1', 'sin/1');
  add_line(modelName, 'Vin/1', 'v*cos/1'); add_line(modelName, 'cos/1', 'v*cos/2');
  add_line(modelName, 'Vin/1', 'v*sin/1'); add_line(modelName, 'sin/1', 'v*sin/2');
  add_line(modelName, 'v*cos/1', 'x/1'); add_line(modelName, 'v*sin/1', 'y/1');
  add_line(modelName, 'Win/1', 'theta/1');
  add_line(modelName, 'x/1', 'x_out/1'); add_line(modelName, 'y/1', 'y_out/1'); add_line(modelName, 'theta/1', 'theta_out/1');

  save_system(modelName);
end
